packages/service-generator-csharp/src/service.ts

```diff
@@ -0,0 +1,831 @@
+import {
+  emitOperationState,
+  emitOperationStatus,
+  emitResourceOperationStatus,
+} from "./boilerplate.js";
+import { CSharpSourceType, CSharpType, ControllerContext, NameCasingType } from "./interfaces.js";
+import { CSharpServiceEmitterOptions } from "./lib.js";
+import {
+  ensureCSharpIdentifier,
+  ensureCleanDirectory,
+  formatComment,
+  getCSharpIdentifier,
+  getCSharpStatusCode,
+  getCSharpTypeForScalar,
+  getOperationVerbDecorator,
+  hasNonMetadataProperties,
+  isValueType,
+} from "./utils.js";
+
+export async function $onEmit(context: EmitContext<CSharpServiceEmitterOptions>) {
+  let _unionCounter: number = 0;
+  const controllers = new Map<string, ControllerContext>();
+  const NoResourceContext: string = "RPCOperations";
+
+  class CSharpCodeEmitter extends CodeTypeEmitter {
+    #names: string[] = [];
+    #licenseHeader = `// Copyright (c) Microsoft Corporation. All rights reserved.
+    // Licensed under the MIT License.`;
+    #sourceTypeKey: string = "sourceType";
+
+    #metaInfo: MetadataInfo = createMetadataInfo(this.emitter.getProgram(), {
+      canonicalVisibility: Visibility.Read,
+      canShareProperty: (p) => true,
+    });
+
+    arrayDeclaration(array: Model, name: string, elementType: Type): EmitterOutput<string> {
+      return this.emitter.result.declaration(
+        ensureCSharpIdentifier(this.emitter.getProgram(), array, name),
+        code`${this.emitter.emitTypeReference(elementType)}[]`
+      );
+    }
+
+    arrayLiteral(array: Model, elementType: Type): EmitterOutput<string> {
+      return this.emitter.result.rawCode(code`${this.emitter.emitTypeReference(elementType)}[]`);
+    }
+
+    booleanLiteral(boolean: BooleanLiteral): EmitterOutput<string> {
+      return this.emitter.result.rawCode(code`${boolean.value === true ? "true" : "false"}`);
+    }
+
+    unionLiteral(union: Union): EmitterOutput<string> {
+      const csType = this.#coalesceUnionTypes(union);
+      return this.emitter.result.rawCode(csType && csType.isBuiltIn ? csType.name : "object");
+    }
+
+    declarationName(declarationType: TypeSpecDeclaration): string {
+      switch (declarationType.kind) {
+        case "Enum":
+        case "Interface":
+        case "Model":
+        case "Operation":
+          return getCSharpIdentifier(declarationType.name, NameCasingType.Class);
+        case "Union":
+          if (!declarationType.name) return `Union${_unionCounter++}`;
+          return declarationType.name;
+        case "Scalar":
+        default:
+          return declarationType.name;
+      }
+    }
+
+    enumDeclaration(en: Enum, name: string): EmitterOutput<string> {
+      const enumName = ensureCSharpIdentifier(this.emitter.getProgram(), en, name);
+      const namespace = this.emitter.getContext().namespace;
+      const doc = getDoc(this.emitter.getProgram(), en);
+      return this.emitter.result.declaration(
+        enumName,
+        code`${this.#licenseHeader}
+        // <auto-generated />
```

Comment:
(Food for thought)

1. I often see a "by XXX" and sometimes a timestamp for generated code. Does this `<auto-generated />` tag get replaced with additional detail or is this a raw comment string?

2. I am still learning C#, but does the [GeneratedCodeAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.codedom.compiler.generatedcodeattribute?view=net-7.0) apply to the use case here? (In Java world, I used `@Generated` IIRC.)

Reply:
The primary reason for this is that they are keys for many c-sharp linting rules